pkg/controllers/corecluster/corecluster.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2024 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package cluster
+
+import (
+	"context"
+	"fmt"
+
+	configv1 "github.com/openshift/api/config/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/openshift/cluster-capi-operator/pkg/operatorstatus"
+	"github.com/openshift/cluster-capi-operator/pkg/util"
+)
+
+const controllerName = "CoreClusterController"
+
+// CoreClusterReconciler reconciles a Cluster object.
+type CoreClusterReconciler struct {
+	operatorstatus.ClusterOperatorStatusClient
+	Cluster  *clusterv1.Cluster
+	Infra    *configv1.Infrastructure
+	Platform configv1.PlatformType
+}
+
+// SetupWithManager sets the CoreClusterReconciler controller up with the given manager.
+func (r *CoreClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := ctrl.NewControllerManagedBy(mgr).
+		Named(controllerName).
+		For(r.Cluster).
+		Complete(r); err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	return nil
+}
+
+// Reconcile reconciles the core cluster object for the openshift-cluster-api namespace.
+func (r *CoreClusterReconciler) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx, "namespace", req.Namespace, "name", req.Name)
+
+	// Still have to patch the condition onto a newly created cluster
+	cluster, err := CreateCoreCluster(r.Client, r.Cluster.Name, r.Infra.Kind)
```

Comment:
I moved the creation into it's own function for readability - but maybe that's unnecessary?

Reply:
That could work, but I see this actually doesn't contain a create call, so it shouldn't probably be called "Create". If you want to actually perform the creation in there then it is probably best to make it a `r.ensureCoreCluster()` similarly to what we did for the `r.ensureInfraCluster()`, but without the extra layer for providers. So basically end up with having the `r.ensureCoreCluster()` have a structure similar to https://github.com/openshift/cluster-capi-operator/blob/ec9120bb11ace966a9046ad8e698216f8abf5c42/pkg/controllers/infracluster/aws.go#L34-L86.

Then later followed by the status patching logic.

Reply:
Let me know what you think of it now? I left the creation out of the ensure function but as you're saying here maybe it's best to keep it together to be tidier